### PR TITLE
DAOS-13006 test: resolve/silence all pylint warnings

### PR DESCRIPTION
**Pull request by daltonbohning**
Test-tag: pr daily_regression SameKeyDifferentValue TestWithTelemetryIOLatency ManagementServiceResilience ManagementServiceFailover soak_smoke HarnessCoreFilesTest test_container_boundary test_pool_management_race test_metadata_fillup test_telemetry_list test_container_telemetry TestWithTelemetryNvme TestHarnessSkipsBase test_scrubber_csum_fault test_scrubber_ssd_auto_eviction test_scrubber_disabled_during_pool_creation test_scrubber_enabled_during_pool_creation test_scrubber_ssd_auto_eviction test_list_pools test_used_imbalance test_fields_basic test_dfuse_find test_get_acl_valid DaosAggregationIOSmall IoSysAdmin container_acl DaosPerfLarge DaosPerf CartRpcOneNodeSwimNotificationOnRankEvictionTest EcodRunIoConf DaosRunIoConf RbldRunIoConf daos_racer DaosServerDumpTest,-manual OverwriteContainerACLTest UpdateContainerACLTest cart TestHarnessSkipsSkipped TestHarnessSkipsRun ParallelIo osa test_snapshot_negativecases test_snapshots test_single_object_validation container test_pool_management_race test_ioaggregation DaosCoreTestNvme DmvrPosixLargeFile TargetFailure EcodCellSizeProperty EcodCellSize EcodDisabledRebuildSingle IorHard OSADmgNegativeTest OSAOnlineDrain PoolAutotestTest
Allow-unstable-test: true

Resolve or silence all remaining pylint issues.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/io/parallel_io.py
Comment (ashleypittman) [inline src/tests/ftest/io/parallel_io.py:74]: From the looks of it.
```suggestion
            statvfs_info = self._stat_free_blocks(dfuse_pool_dir)
```

File: src/tests/ftest/object/integrity.py
Comment (ashleypittman) [inline src/tests/ftest/object/integrity.py:1]: As above.

File: src/tests/ftest/util/general_utils.py
Comment (ashleypittman) [inline src/tests/ftest/util/general_utils.py:212]: lolz.

File: src/tests/ftest/util/job_manager_utils.py
Comment (ashleypittman) [inline src/tests/ftest/util/job_manager_utils.py:8]: This is a new warning, and I think we use it in the build as well, we should file a ticket to remove this.

File: src/tests/ftest/util/job_manager_utils.py
Comment (ashleypittman) [inline src/tests/ftest/util/job_manager_utils.py:376]: None is not a bool

File: src/tests/ftest/util/job_manager_utils.py
Comment (ashleypittman) [inline src/tests/ftest/util/job_manager_utils.py:382]: Still not a bool.

File: utils/cq/words.dict
Comment (ashleypittman) [inline utils/cq/words.dict:325]: Can you remove fnctl from this file, it's a spelling mistake itself.

**Review by ashleypittman — COMMENTED** [review]
There is logic in daos_pylint.py that will fail the GHA job for any errors in scons code, not just modified scons code, if ftest is clean with this PR then the same logic should be applied there - it will mean that any test escapes get picked up on future PRs and landings builds rather than ignored as not regressions.

You need to update some code close to here to enable that, it'll cause a non-zero exit failure which will make the pylint GHA job fail/red which should in turn mean PRs cannot be landed without force landing.
https://github.com/daos-stack/daos/blob/master/utils/cq/daos_pylint.py#L436-L438

File: src/tests/ftest/deployment/critical_integration.py
Comment (daltonbohning) [reply src/tests/ftest/deployment/critical_integration.py:19]: Agree. Thanks

File: src/tests/ftest/object/fetch_bad_param.py
Comment (daltonbohning) [reply src/tests/ftest/object/fetch_bad_param.py:1]: The file has `X` set, but it should NOT. Being resolved by #11812

File: src/tests/ftest/util/job_manager_utils.py
Comment (daltonbohning) [reply src/tests/ftest/util/job_manager_utils.py:376]: Right. But when something IS passed, it's a `bool`. I would think the default here should be `False`, but out of the scope of this PR

**Comment by daltonbohning** [comment]
> There is logic in daos_pylint.py that will fail the GHA job for any errors in scons code, not just modified scons code, if ftest is clean with this PR then the same logic should be applied there - it will mean that any test escapes get picked up on future PRs and landings builds rather than ignored as not regressions.
> 
> You need to update some code close to here to enable that, it'll cause a non-zero exit failure which will make the pylint GHA job fail/red which should in turn mean PRs cannot be landed without force landing. https://github.com/daos-stack/daos/blob/master/utils/cq/daos_pylint.py#L436-L438

Thanks. Done. Once #11812 lands, I'll need to rebase this PR (there is some overlap)

File: src/tests/ftest/control/ms_resilience.py
Comment (ashleypittman) [inline src/tests/ftest/control/ms_resilience.py:205]: Any idea why we don't need a "nosec" for this random call?

File: utils/githooks/prepare-commit-msg.d/10-submodule-update-check.py
Comment (ashleypittman) [inline utils/githooks/prepare-commit-msg.d/10-submodule-update-check.py:1]: Could possibly just rename this file.

**Review by ashleypittman — APPROVED** [review]

File: src/tests/ftest/util/job_manager_utils.py
Comment (daltonbohning) [reply src/tests/ftest/util/job_manager_utils.py:8]: https://daosio.atlassian.net/browse/DAOS-13130

File: src/tests/ftest/control/ms_resilience.py
Comment (daltonbohning) [reply src/tests/ftest/control/ms_resilience.py:205]: Apparently, `random.sample` isn't on the blacklist:
# pylint: disable-next=relative-beyond-top-level

File: utils/githooks/prepare-commit-msg.d/10-submodule-update-check.py
Comment (daltonbohning) [reply utils/githooks/prepare-commit-msg.d/10-submodule-update-check.py:1]: The problem is that `10_submodule_update_check.py` is also invalid, because it starts with a number

**Comment by daltonbohning** [comment]
@ashleypittman How do you feel about my latest commit that ignores remaining pylint? The idea is to get to a "clean" slate so we can enforce pylint on PRs. Then we can worry about cleaning up what we're currently ignoring

**Comment by ashleypittman** [comment]
> @ashleypittman How do you feel about my latest commit that ignores remaining pylint? The idea is to get to a "clean" slate so we can enforce pylint on PRs. Then we can worry about cleaning up what we're currently ignoring

The ftest ones and the githooks look OK but the pydaos is possibly a little heavy-handed.

If pylint is clean however then we should also update the commit hook, https://github.com/daos-stack/daos/blob/master/.github/workflows/pylint.yml#L26-L28 would need updating to not pass in a diff so it would treat all errors equally, and https://github.com/daos-stack/daos/blob/master/utils/cq/daos_pylint.py#L640-L662 and the FileTypeList could be simplified to only take the filename from the diff rather than the regions as well for the commit hook.

Then you hit a potential issue where the commit hook itself is run from the target but possibly using the code from the branch (you'd need to check which) so the above would need to be coordinated to ensure compatibility (I've not done this before and it's caught us out).

If you land this as-is I'd expect the code to be clean for a very short period of time but a regression would get landed or force-landed and we might not notice it, so I'd prefer to keep this as a "big clean-up", with then a second smaller PR going in to catch the stragglers, and have the implications of that propagated to the gatekeepers before we land.  If we have any feature branches merging in the near future then we'll want to ask them to resolve issues as well (they should be anyway but I doubt we've checked).

**Comment by daltonbohning** [comment]
> The ftest ones and the githooks look OK but the pydaos is possibly a little heavy-handed.

Do you think I should backout the pydaos ignore? Specifically, the `invalid-name` ignores the most.


> If pylint is clean however then we should also update the commit hook, https://github.com/daos-stack/daos/blob/master/.github/workflows/pylint.yml#L26-L28 would need updating to not pass in a diff so it would treat all errors equally, and https://github.com/daos-stack/daos/blob/master/utils/cq/daos_pylint.py#L640-L662 and the FileTypeList could be simplified to only take the filename from the diff rather than the regions as well for the commit hook.
> 
> Then you hit a potential issue where the commit hook itself is run from the target but possibly using the code from the branch (you'd need to check which) so the above would need to be coordinated to ensure compatibility (I've not done this before and it's caught us out).
> 
> If you land this as-is I'd expect the code to be clean for a very short period of time but a regression would get landed or force-landed and we might not notice it, so I'd prefer to keep this as a "big clean-up", with then a second smaller PR going in to catch the stragglers, and have the implications of that propagated to the gatekeepers before we land. If we have any feature branches merging in the near future then we'll want to ask them to resolve issues as well (they should be anyway but I doubt we've checked).

In general, I've seen some slightly different behavior from the pylint check when ran from the gitook vs executing `daos_pylint.py` directly vs the GHA check. So I'm hesitant to change the behavior of `daos_pylint.py` too much in this PR, in case there's something I've forgotten. And you bring up a good point about feature branches, since they likely have more linting issues. @phender and I talked about this some, and I wonder if we should (1) land this big cleanup PR, (2) notify devs, feature branch owners, and gatekeepers about the "clean slate", (3) give everyone time to catch up, (4) resolve remaining issues, simplify lint checks, and tighten the checks.
Thoughts @ashleypittman?

**Comment by ashleypittman** [comment]
> > The ftest ones and the githooks look OK but the pydaos is possibly a little heavy-handed.
> 
> Do you think I should backout the pydaos ignore? Specifically, the `invalid-name` ignores the most.
> 
> > If pylint is clean however then we should also update the commit hook, https://github.com/daos-stack/daos/blob/master/.github/workflows/pylint.yml#L26-L28 would need updating to not pass in a diff so it would treat all errors equally, and https://github.com/daos-stack/daos/blob/master/utils/cq/daos_pylint.py#L640-L662 and the FileTypeList could be simplified to only take the filename from the diff rather than the regions as well for the commit hook.
> > Then you hit a potential issue where the commit hook itself is run from the target but possibly using the code from the branch (you'd need to check which) so the above would need to be coordinated to ensure compatibility (I've not done this before and it's caught us out).
> > If you land this as-is I'd expect the code to be clean for a very short period of time but a regression would get landed or force-landed and we might not notice it, so I'd prefer to keep this as a "big clean-up", with then a second smaller PR going in to catch the stragglers, and have the implications of that propagated to the gatekeepers before we land. If we have any feature branches merging in the near future then we'll want to ask them to resolve issues as well (they should be anyway but I doubt we've checked).
> 
> In general, I've seen some slightly different behavior from the pylint check when ran from the gitook vs executing `daos_pylint.py` directly vs the GHA check. So I'm hesitant to change the behavior of `daos_pylint.py` too much in this PR, in case there's something I've forgotten. And you bring up a good point about feature branches, since they likely have more linting issues. @phender and I talked about this some, and I wonder if we should (1) land this big cleanup PR, (2) notify devs, feature branch owners, and gatekeepers about the "clean slate", (3) give everyone time to catch up, (4) resolve remaining issues, simplify lint checks, and tighten the checks. Thoughts @ashleypittman?

Well, the behaviour is supposed to be different.  I wasn't suggesting doing that in this PR, but perhaps we want to save the last few fixes and the tidy up work there for a second PR that as you say would need more process in place about landing it.

File: src/client/pydaos/pydaos_core.py
Comment (daltonbohning) [reply src/client/pydaos/pydaos_core.py:6]: I agree we shouldn't use python2 - it was EOL January 1, 2020. But I was trying to avoid any breaking changes

File: src/client/pydaos/__init__.py
Comment (daltonbohning) [reply src/client/pydaos/__init__.py:14]: Done

**Comment by daltonbohning** [comment]
@ashleypittman I should mention I haven't ran any tests yet for this PR, so making changes is trivial. Once we're set, I'll run the tests (it's a lot)

File: src/client/pydaos/__init__.py
Comment (daltonbohning) [inline src/client/pydaos/__init__.py:25]: @ashleypittman FYI - This one just came up locally, after not seeing it before. And this file hasn't been modified since I last merged master. So I wonder if more like this exist that just aren't being flagged

File: utils/cq/daos_pylint.py
Comment (ashleypittman) [inline utils/cq/daos_pylint.py:437]: This isn't technically correct now but we can fix it up later when we remove no-longer-needed code here.

File: src/client/pydaos/pydaos_core.py
Comment (ravalsam) [inline src/client/pydaos/pydaos_core.py:1]: This file still has some pylint warning. So work is still in progress ?

`
************* Module pydaos.pydaos_core
src/client/pydaos/pydaos_core.py:6:0: E0012: Bad option value 'consider-using-f-string' (bad-option-value)
src/client/pydaos/pydaos_core.py:13:0: E0011: Unrecognized file option 'disable-next' (unrecognized-inline-option)
src/client/pydaos/pydaos_core.py:195:0: E0012: Bad option value 'unnecessary-dunder-call' (bad-option-value)
src/client/pydaos/pydaos_core.py:155:4: R0913: Too many arguments (6/5) (too-many-arguments)

------------------------------------------------------------------
Your code has been rated at 9.18/10 (previous run: 9.15/10, +0.03)

`

Comment (daltonbohning) [reply src/client/pydaos/pydaos_core.py:1]: @ravalsam Where are you seeing that?

Comment (ravalsam) [reply src/client/pydaos/pydaos_core.py:1]: On my local machine

Comment (ashleypittman) [reply src/client/pydaos/pydaos_core.py:1]: if your pylint up-to-date?

It's best to run ./utils/cq/daos_pylint as that will apply the correct settings for the daos project.

Comment (ravalsam) [reply src/client/pydaos/pydaos_core.py:1]: Yes, I was trying with pylint-3 command. I tried running with daos_pylint.py and it looks good.

File: site_scons/env_modules.py
Comment (ashleypittman) [inline site_scons/env_modules.py:66]: I think this was to get scons running on python2 and 3, it could probably be simplified now, but outside the scope of this PR.

**Comment by ashleypittman** [comment]
There are some good fixes in this PR and it touches a large number of files, it would be good to prioritise getting this in.

**Comment by daltonbohning** [comment]
> There are some good fixes in this PR and it touches a large number of files, it would be good to prioritise getting this in.

I've been trying, but with unrelated test failures and CI instability, it's incredibly slow progress.

**Comment by daltonbohning** [comment]
Failures in https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-11825/34/tests
- daos_racer/parallel.py - https://daosio.atlassian.net/browse/DAOS-12828
- erformance/ior_hard.py - manual tests that I shouldn't have ran

File: src/client/pydaos/pydaos_core.py
Comment (phender) [inline src/client/pydaos/pydaos_core.py:199]: We should keep this newline.

**Review by phender — APPROVED** [review]

File: src/tests/ftest/config_file_gen.py
Comment (ashleypittman) [reply src/tests/ftest/config_file_gen.py:7]: As I put above there is something odd about how this file is run and where it lives in the source tree, I think in order to run it it would either have to be copied to a different location or else run with a custom PYTHONPATH as the `from utli.*` import isn't working in the checker.  This means that some checks that pylint could run are not being run so we've not getting full coverage, but it's also a bad code smell that (I think) is indicating that our source code is laid out incorrectly.

I agree that we should move these annotations to the exact lines that require them and use disable-next to limit the scope of what's being disabled but I don't think that should block the PR.

There is an issue with this code however that's "odd" and we've already taken considerable time trying to unwind it and find out the root cause of the problem, the TODO could certainly be expanded to better explain our current understanding (vague such as it is) but to remove the marker noting that there are problems here would be a mistake.  From recollection there were 3-4 similar issues across the codebase when I worked on adding the pylint testing, the others I was able to resolve adequately but this one which affects not just this file but at least one more are still open questions.

Comment (daltonbohning) [reply src/tests/ftest/config_file_gen.py:7]: @ashleypittman Talked with Phil and we're okay to look into / fix this later

File: src/tests/ftest/util/apricot/apricot/test.py
Comment (ashleypittman) [reply src/tests/ftest/util/apricot/apricot/test.py:1330]: I dislike disabling the arguments-renamed warning, this to me seems like the kind of subtle issue we are looking for personally.

File: src/tests/ftest/util/data_mover_test_base.py
Comment (ashleypittman) [inline src/tests/ftest/util/data_mover_test_base.py:740]: I think this was probably correct, at least according to the methods.  Frankly I'm far from sure though.

File: utils/ansible/ftest/library/daos_hugepages.py
Comment (ashleypittman) [inline utils/ansible/ftest/library/daos_hugepages.py:1]: Be good to fix rather than suppress the issues in this file but not a regression.

**Review by ashleypittman — APPROVED** [review]

**Review by ashleypittman — APPROVED** [review]

**Review by ashleypittman — APPROVED** [review]

**Comment by daltonbohning** [comment]
Previous run passed except https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-11825/34/tests
- daos_racer/parallel.py - https://daosio.atlassian.net/browse/DAOS-12828
- performance/ior_hard.py - manual tests that I shouldn't have ran

Latest run after merge conflict resolution ran just `soak_smoke` and it passed
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-11825/37/tests/

**Pull request merged**
